Mathlib/Algebra/Ring/Equiv.lean

```diff
@@ -864,24 +865,34 @@ theorem symm_trans_self (e : R ≃+* S) : e.symm.trans e = RingEquiv.refl S :=
   ext e.right_inv
 #align ring_equiv.symm_trans_self RingEquiv.symm_trans_self
 
+end RingEquiv
+
+namespace MulEquiv
+
 /-- If two rings are isomorphic, and the second doesn't have zero divisors,
 then so does the first. -/
-protected theorem noZeroDivisors {A : Type*} (B : Type*) [Ring A] [Ring B] [NoZeroDivisors B]
-    (e : A ≃+* B) : NoZeroDivisors A :=
-  { eq_zero_or_eq_zero_of_mul_eq_zero := fun {x y} hxy => by
-      have : e x * e y = 0 := by rw [← e.map_mul, hxy, e.map_zero]
-      simpa using eq_zero_or_eq_zero_of_mul_eq_zero this }
-#align ring_equiv.no_zero_divisors RingEquiv.noZeroDivisors
+protected theorem noZeroDivisors {A : Type*} (B : Type*) [MulZeroClass A] [MulZeroClass B]
+    [NoZeroDivisors B] (e : A ≃* B) : NoZeroDivisors A :=
+  e.injective.noZeroDivisors e (map_zero e) (map_mul e)
+#noalign ring_equiv.no_zero_divisors
 
 /-- If two rings are isomorphic, and the second is a domain, then so is the first. -/
-protected theorem isDomain {A : Type*} (B : Type*) [Ring A] [Ring B] [IsDomain B] (e : A ≃+* B) :
-    IsDomain A := by
-  haveI : Nontrivial A := ⟨⟨e.symm 0, e.symm 1, e.symm.injective.ne zero_ne_one⟩⟩
-  haveI := e.noZeroDivisors B
-  exact NoZeroDivisors.to_isDomain _
-#align ring_equiv.is_domain RingEquiv.isDomain
+protected theorem isDomain {A : Type*} (B : Type*) [Semiring A] [Semiring B] [IsDomain B]
+    (e : A ≃* B) : IsDomain A :=
+  { e.injective.isLeftCancelMulZero e (map_zero e) (map_mul e),
+    e.injective.isRightCancelMulZero e (map_zero e) (map_mul e) with
+    exists_pair_ne := ⟨e.symm 0, e.symm 1, e.symm.injective.ne zero_ne_one⟩ }
+#noalign ring_equiv.is_domain
+
+protected theorem isField {A : Type*} (B : Type*) [Semiring A] [Semiring B] (hB : IsField B)
+    (e : A ≃* B) : IsField A where
```

Comment:
Is `IsField` actually used or is it mostly deprecated? That's a genuine question, I didn't even know it existed :sweat_smile:

Reply:
It's used [pretty often](https://github.com/search?q=repo%3Aleanprover-community%2Fmathlib4%20isfield&type=code). I guess if you have a nontrivial CommRing to start with, this is the only way to say that every nonzero element is invertible (i.e. it's a field).

Reply:
Oh right that makes sense. I don't use this part of the library very often so it's not that surprising that I missed it.